angr/analyses/find_objects_static.py

```diff
@@ -0,0 +1,130 @@
+                              codeloc=None):
+        if function_address == self.new_func_addr:
+            # check if this is a call to new()
+            # reading from rdi for the size argument passed to new()
+            size = state.register_definitions.load(72, state.arch.bits//state.arch.byte_width).values[0].pop()._model_concrete.value
+
+            # None since we do not know it's class yet, it a possible this pointer
+            self.possible_objects_dict[self.max_addr] = PossibleObject(size, self.max_addr)
+
+            # assigning eax a concrete address to track the possible this pointer
+            atom = Register(16, 8)
+            state.kill_and_add_definition(atom, codeloc, MultiValues(offset_to_values={0: {claripy.BVV(self.max_addr,
+                                     8 * state.arch.byte_width
+                                )}}))
+            # setting the values pointed by rax to zero
+            memory_location = MemoryLocation(self.max_addr, size)
+            offset_to_values = {}
+            for offset in range(0, size, 8):
+                offset_to_values[offset] = {claripy.BVV(0, 8 * state.arch.byte_width)}
+            state.kill_and_add_definition(memory_location, codeloc, MultiValues(offset_to_values=offset_to_values))
+            self.max_addr += size
+
+        elif "ctor" in self.cfg.kb.functions[function_address].demangled_name:
+            # check if rdi has a possible this pointer/ object address, if so then we can assign this object this class
+            # also if the func is a constructor(not stripped binaries)
+            for addr in self.possible_objects_dict:
+                obj_addr = state.register_definitions.load(72, state.arch.bits // state.arch.byte_width).values[0].pop()._model_concrete.value
+                if addr == obj_addr:
+                    col_ind = self.cfg.kb.functions[function_address].demangled_name.rfind("::")
+                    class_name = self.cfg.kb.functions[function_address].demangled_name[:col_ind]
+                    self.possible_objects_dict[addr].class_name = class_name
+        executed_rda = True
+        return executed_rda, state, visited_blocks, dep_graph
+
+
+class StaticObjectFinder(ForwardAnalysis, Analysis):
+    def __init__(self, func=None, func_graph=None, block=None, max_iterations=1):
+        vtable_analysis = self.project.analyses.VtableFinder()
+        self.vtables_list = vtable_analysis.vtables_list
+        self.possible_objects = set()
+
+        # for stripped binaries
+        # This is a set of functions that are possible constructors
+        self.possible_constructors = set()
+
+        self._analyze()
+
+    def is_new_func(self, state):
+        if state.solver.eval(state.inspect.function_address) in self.possible_new_functions:
+            return True
+        else:
+            False
+
+    def _analyze(self):
+        self.cfg = self.project.analyses.CFGFast(cross_references=True)

```

Comment:
Can we make this use an existing cfg if one already is present in the knowledge base?